### ISSUE TITLE
Second LIST of the same symlink shows different results

### ISSUE THREAD

**Issue by dimsua** [proftpd/proftpd#940]
### What I Did
Hello
When logging and make first ls symlink list ok test2 -> test
```
ftp> ls data
227 Entering Passive Mode (95,183,14,53,234,127).
150 Opening ASCII mode data connection for file list
drwxr-xr-x   4 p284174  284174       4096 Mar 19 09:15 .
drwxr-x---  14 99       284174       4096 Apr 26  2019 ..
drwxrwxr-x   3 p284174  284174       4096 Mar 19 09:15 old
-rw-rw-r--   1 p284174  284174          0 Oct 18  2017 .pearrc
drwxrwxr-x   2 p284174  284174       4096 Mar 16 10:25 test
lrwxrwxrwx   1 p284174  284174          4 Mar 16 10:25 test2 -> test
226 Transfer complete
```
But when make second list after first and other listing after test2 listing as dir and not symlink
```
ftp> ls data
227 Entering Passive Mode (95,183,14,53,198,170).
150 Opening ASCII mode data connection for file list
drwxr-xr-x   4 p284174  284174       4096 Mar 19 09:15 .
drwxr-x---  14 99       284174       4096 Apr 26  2019 ..
drwxrwxr-x   3 p284174  284174       4096 Mar 19 09:15 old
-rw-rw-r--   1 p284174  284174          0 Oct 18  2017 .pearrc
drwxrwxr-x   2 p284174  284174       4096 Mar 16 10:25 test
drwxrwxr-x   2 p284174  284174       4096 Mar 16 10:25 test2
226 Transfer complete
```

### What I Expected/Wanted

We want that symlink always listed as symlink, not a dir

### ProFTPD Version and Configuration

```
proftpd -V
Compile-time Settings:
  Version: 1.3.6c (maint)
  Platform: LINUX [Linux 3.10.0-962.3.2.lve1.5.26.1.el7.x86_64 x86_64]
  Built: Thu Mar 19 2020 12:07:42 MSK
  Built With:
    configure  '--build=x86_64-redhat-linux-gnu' '--host=x86_64-redhat-linux-gnu' '--program-prefix=' '--disable-dependency-tracking' '--prefix=/usr' '--exec-prefix=/usr' '--bindir=/usr/bin' '--sbindir=/usr/sbin' '--sysconfdir=/etc' '--datadir=/usr/share' '--includedir=/usr/include' '--libdir=/usr/lib64' '--libexecdir=/usr/libexec' '--localstatedir=/var' '--sharedstatedir=/var/lib' '--mandir=/usr/share/man' '--infodir=/usr/share/info' '--libexecdir=/usr/libexec/proftpd' '--localstatedir=/var/run/proftpd' '--disable-strip' '--enable-ctrls' '--enable-dso' '--enable-facl' '--enable-ipv6' '--enable-nls' '--enable-openssl' '--enable-shadow' '--with-libraries=/usr/lib64/mysql' '--with-includes=/usr/include/mysql' '--with-modules=mod_readme:mod_auth_pam:mod_tls:mod_vroot:mod_copy:mod_clamav:mod_fsync:mod_tls_redis:mod_tls_memcache' '--with-shared=mod_sql:mod_sql_passwd:mod_sql_mysql:mod_sql_postgres:mod_quotatab:mod_quotatab_file:mod_quotatab_ldap:mod_quotatab_radius:mod_quotatab_sql:mod_ldap:mod_ban:mod_wrap:mod_ctrls_admin:mod_facl:mod_load:mod_radius:mod_ratio:mod_rewrite:mod_site_misc:mod_exec:mod_shaper:mod_geoip:mod_wrap2:mod_wrap2_file:mod_wrap2_sql:mod_sftp:mod_sftp_pam:mod_sftp_sql:mod_tls_shmcache:mod_ifsession' 'build_alias=x86_64-redhat-linux-gnu' 'host_alias=x86_64-redhat-linux-gnu' 'CFLAGS=-O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector-strong --param=ssp-buffer-size=4 -grecord-gcc-switches   -m64 -mtune=generic' 'LDFLAGS=-Wl,-z,relro ' 'CXXFLAGS=-O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector-strong --param=ssp-buffer-size=4 -grecord-gcc-switches   -m64 -mtune=generic'

  CFLAGS: -g2 -O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector-strong --param=ssp-buffer-size=4 -grecord-gcc-switches   -m64 -mtune=generic -Wall -fno-omit-frame-pointer -Werror=implicit-function-declaration
  LDFLAGS: -L$(top_srcdir)/lib -Wl,-z,relro  -rdynamic -L/usr/lib64/mysql -L/usr/lib64/mysql -L/usr/lib64
  LIBS: -lacl  -lssl -lcrypto -lcap  -lssl -lcrypto  -lpam -lsupp -lattr -lnsl -lresolv -lresolv -lcrypt -ldl

  Files:
    Configuration File:
      /etc/proftpd.conf
    Pid File:
      /var/run/proftpd/proftpd.pid
    Scoreboard File:
      /var/run/proftpd/proftpd.scoreboard
    Header Directory:
      /usr/include/proftpd
    Shared Module Directory:
      /usr/libexec/proftpd

  Info:
    + Max supported UID: 4294967295
    + Max supported GID: 4294967295

  Features:
    - Autoshadow support
    + Controls support
    + curses support
    - Developer support
    + DSO support
    + IPv6 support
    + Largefile support
    - Lastlog support
    - Memcache support
    + ncursesw support
    + NLS support
    - Redis support
    - Sodium support
    + OpenSSL support (FIPS enabled)
    - PCRE support
    + POSIX ACL support
    + Shadow file support
    + Sendfile support
    + Trace support
    + xattr support

  Tunable Options:
    PR_TUNABLE_BUFFER_SIZE = 1024
    PR_TUNABLE_DEFAULT_RCVBUFSZ = 8192
    PR_TUNABLE_DEFAULT_SNDBUFSZ = 8192
    PR_TUNABLE_ENV_MAX = 2048
    PR_TUNABLE_GLOBBING_MAX_MATCHES = 100000
    PR_TUNABLE_GLOBBING_MAX_RECURSION = 8
    PR_TUNABLE_HASH_TABLE_SIZE = 40
    PR_TUNABLE_LOGIN_MAX = 256
    PR_TUNABLE_NEW_POOL_SIZE = 512
    PR_TUNABLE_PATH_MAX = 4096
    PR_TUNABLE_SCOREBOARD_BUFFER_SIZE = 80
    PR_TUNABLE_SCOREBOARD_SCRUB_TIMER = 30
    PR_TUNABLE_SELECT_TIMEOUT = 30
    PR_TUNABLE_TIMEOUTIDENT = 10
    PR_TUNABLE_TIMEOUTIDLE = 600
    PR_TUNABLE_TIMEOUTLINGER = 10
    PR_TUNABLE_TIMEOUTLOGIN = 300
    PR_TUNABLE_TIMEOUTNOXFER = 300
    PR_TUNABLE_TIMEOUTSTALLED = 3600
    PR_TUNABLE_XFER_SCOREBOARD_UPDATES = 10

```

proftpd.conf
```
ServerName                      "ProFTPD server"
ServerIdent                     on "FTP Server ready."
ServerAdmin                     support@localhost
ServerType                      standalone
DefaultServer                   on
AccessGrantMsg                  "User %u logged in."
DeferWelcome                    off
DefaultRoot                     ~ !adm
AuthPAMConfig                    proftpd
AuthOrder mod_auth_file.c
AuthUserFile /etc/proftpd.passwd
RequireValidShell off
IdentLookups                    off
UseReverseDNS                   off
Port                            21
PassivePorts			49152 65534
Umask                           022
MaxClientsPerHost               10
SetEnv TZ :/etc/localtime
AllowRetrieveRestart            on
AllowStoreRestart               on
MaxInstances                    60
MaxConnectionsPerHost           5 "Sorry, you may not connect more"
User                            nobody
Group                           nobody
UseSendfile                     on
<Global>
  AllowOverwrite                yes
  <Limit ALL SITE_CHMOD>
    AllowAll
  </Limit>
</Global>
LogFormat                       default "%h %l %u %t \"%r\" %s %b"
LogFormat                       auth    "%v [%P] %h %t \"%r\" %s"
SyslogLevel info
SystemLog    /var/log/proftpd/system.log     ALL default
<IfModule mod_cap.c>
        CapabilitiesEngine      off
</IfModule>
<IfModule mod_clamav.c>
    ClamAV off
    ClamServer 127.0.0.1
    ClamPort 3310
    ClamMaxSize 1 Mb
</IfModule>
<IfModule mod_fsync.c>
    FsyncEngine off
</IfModule>
RLimitCPU session 100
<IfModule mod_ban.c>
  BanEngine on
  BanLog /var/log/proftpd/ban.log
  BanTable /var/run/proftpd/ban.tab
  BanOnEvent MaxLoginAttempts 2/00:10:00 01:00:00
  BanControlsACLs all allow user ftpadm
</IfModule>
  <IfModule mod_tls.c>
    TLSEngine on
    TLSLog /var/log/proftpd/tls.log
    TLSProtocol TLSv1 TLSv1.1 TLSv1.2
    TLSRequired off
    TLSCipherSuite EECDH:+AES256:-3DES:RSA+AES:RSA+3DES:!NULL:!RC4
    TLSRSACertificateFile /etc/proftpd.d/cert/proftpd.crt
    TLSRSACertificateKeyFile /etc/proftpd.d/cert/proftpd.key
    TLSDHParamFile /etc/proftpd.d/cert/dh1024.pem
    TLSCertificateChainFile /etc/proftpd.d/cert/proftpd-ca.crt
    TLSVerifyClient off
    TLSRenegotiate none
    TLSOptions NoSessionReuseRequired
    TLSSessionCache internal: 1800
  </IfModule>
DefaultRoot ~
<Directory /home/*>
HideFiles (\.cagefs|\.cl\.selector)$
       <Limit ALL>
       IgnoreHidden on
       </Limit>
</Directory>
ListOptions "-la"
```

**Comment by dimsua** [comment]
In strace
```
lstat("test", {st_mode=S_IFDIR|0775, st_size=4096, ...}) = 0
lstat("/data/test", {st_mode=S_IFDIR|0775, st_size=4096, ...}) = 0
stat("/data/test", {st_mode=S_IFDIR|0775, st_size=4096, ...}) = 0
stat("/data/test/.ftpaccess", 0x7ffc21028d80) = -1 ENOENT (No such file or directory)
lstat("test2", {st_mode=S_IFLNK|0777, st_size=4, ...}) = 0
readlink("test2", "test", 4094)         = 4
lstat("/data/test2", {st_mode=S_IFLNK|0777, st_size=4, ...}) = 0
stat("/data/test2", {st_mode=S_IFDIR|0775, st_size=4096, ...}) = 0
stat("/data/test2/.ftpaccess", 0x7ffc21028d80) = -1 ENOENT (No such file or directory)
lstat("test2", {st_mode=S_IFLNK|0777, st_size=4, ...}) = 0
readlink("test2", "test", 4094)         = 4
readlink("test2", "test", 4096)         = 4
write(14, "drwxr-xr-x   4 p284174  284174       4096 Mar 19 09:15 .\r\ndrwxr-x---  14 99       284174       4096 Apr 26  2019 ..\r\ndrwxrwxr-x   3 p284174  284174       4096 Mar 19 09:15 old\r\n-rw-rw-r--   1 p284174  284174          0 Oct 18  2017 .pearrc\r\ndrwxrwxr-x   2 p284174  284174       4096 Mar 16 10:25 test\r\nlrwxrwxrwx   1 p284174  284174          4 Mar 16 10:25 test2 -> test\r\n", 372) = 372
lstat("test", {st_mode=S_IFDIR|0775, st_size=4096, ...}) = 0
lstat("/data/test", {st_mode=S_IFDIR|0775, st_size=4096, ...}) = 0
stat("/data/test", {st_mode=S_IFDIR|0775, st_size=4096, ...}) = 0
lstat("test2", {st_mode=S_IFLNK|0777, st_size=4, ...}) = 0
readlink("test2", "test", 4094)         = 4
lstat("/data/test2", {st_mode=S_IFLNK|0777, st_size=4, ...}) = 0
stat("test2", {st_mode=S_IFDIR|0775, st_size=4096, ...}) = 0
stat("/data/test2", {st_mode=S_IFDIR|0775, st_size=4096, ...}) = 0
readlink("/data/test2", "test", 4096)   = 4
lstat("test2", {st_mode=S_IFLNK|0777, st_size=4, ...}) = 0
readlink("test2", "test", 4094)         = 4
readlink("test2", "test", 4096)         = 4
write(14, "drwxr-xr-x   4 p284174  284174       4096 Mar 19 09:15 .\r\ndrwxr-x---  14 99       284174       4096 Apr 26  2019 ..\r\ndrwxrwxr-x   3 p284174  284174       4096 Mar 19 09:15 old\r\n-rw-rw-r--   1 p284174  284174          0 Oct 18  2017 .pearrc\r\ndrwxrwxr-x   2 p284174  284174       4096 Mar 16 10:25 test\r\ndrwxrwxr-x   2 p284174  284174       4096 Mar 16 10:25 test2\r\n", 364) = 364
```

**Comment by Castaglia** [comment]
Is your FTP client using the `LIST` command, or `MLSD`, for these directory listings?

**Comment by dimsua** [comment]
we are testing with default ftp-console client in linux ftp-0.17-67.el7.x86_64
in strace-log i have see only LIST command no MLSD
read(0, "LIST data\r\n", 4102)          = 11
read(0, "LIST data\r\n", 4102)          = 11

**Comment by Castaglia** [comment]
I'm working to reproduce this behavior locally.  While doing so, I just wanted to point out that the provided `proftpd.conf` (thanks for that, by the way!) has duplicate `DefaultRoot` directives:
```
DefaultRoot                     ~ !adm
...
DefaultRoot ~
```
I'm not sure if that is intentional..

**Comment by Castaglia** [comment]
This has now been fixed in `master`, and in the 1.3.6 branch.  Thanks!

**Comment by dimsua** [comment]
Thank you!
